### ISSUE TITLE
zipkin starts a new traceID

### ISSUE THREAD

**Issue by trajano** [containous/traefik#5511]
# Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [X] No

### What did you do?

I created an NGINX proxy in front of Traefik that uses the opentracing module.

```
load_module modules/ngx_http_opentracing_module.so;
user nginx;
worker_processes 1;
error_log /var/log/nginx/error.log warn;
pid /var/run/nginx.pid;
events {
    worker_connections 1024;
}
http {
    opentracing on;
    opentracing_load_tracer /usr/local/lib/libzipkin_opentracing_plugin.so /etc/zipkin-config.json;
    default_type application/octet-stream;
    log_format main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';
    access_log /var/log/nginx/access.log  main;
    server {
        listen 443;
        ssl on;
        server_name i.trajano.net;
        ssl_certificate /cert.pem;
        ssl_certificate_key /key.pem;
        ssl_client_certificate /ca.pem;
        ssl_verify_client on;
        ssl_trusted_certificate /cert.pem;
        ssl_protocols TLSv1.3 TLSv1.2 TLSv1.1 TLSv1;
        ssl_prefer_server_ciphers on;
        ssl_ciphers EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA512:EECDH+ECDSA+SHA384:EECDH+ECDSA+SHA256:ECDH+AESGCM:ECDH+AES256:DH+AESGCM:DH+AES256:RSA+AESGCM:!aNULL:!eNULL:!LOW:!RC4:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS;
        location / {
            # opentracing_trace_locations off;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_pass http://proxy/;
        }
    }
}
```

I can verify in zipkin that a trace is created by nginx.  However, when I try to access a service that is managed by traefik, it appears to ignore nginx from the trace.

Example (requires certs to run) in https://github.com/trajano/trajano-swarm noted in https://github.com/trajano/trajano-swarm/commit/d18d1edb666072484c98d08fd0e32f5111cfa0f3

### What did you expect to see?

Nginx then traefik as a tree in zipkin

### What did you see instead?

Only traefik

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
v2.0.0
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```
  proxy:
    image: traefik:v2.0.0
    ports:
      - 55555:8080
    command:
      - --global.checknewversion=false
      - --global.sendanonymoususage=false

      - --accesslog
      - --api
      - --api.insecure
      - --ping

      - --entrypoints.http.address=:80

      - --providers.docker.endpoint=tcp://daemon:2375
      - --providers.docker.exposedByDefault=false
      - --providers.docker.swarmMode=true
      - --providers.docker.network=intranet
      - "--providers.docker.constraints=Label(`intranet`, `true`)"
      - '--providers.docker.defaultrule=Host(`i.trajano.net`) && PathPrefix(`/{{ .Name | trimPrefix (index .Labels "com.docker.stack.namespace") | trimPrefix "_" }}`)'

      - --providers.file.filename=/traefik-conf.toml

      - --tracing.zipkin.httpEndpoint=http://zipkin:9411/api/v2/spans
    healthcheck:
      test: traefik healthcheck --ping
    deploy:
      replicas: 1
      update_config:
        order: start-first
      restart_policy:
        condition: any
        delay: 10s
    environment:
      - TZ=America/Toronto
    volumes:
      - /d/p/trajano.net/traefik.toml:/traefik-conf.toml
    networks:
      - intranet
      - internal
      - management
```

config is not really relevant but here you go
```toml
[tls.options]

  [tls.options.default]
    minVersion = "VersionTLS12"

    cipherSuites = [
        "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
        "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
        "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
        "TLS_AES_128_GCM_SHA256",
        "TLS_AES_256_GCM_SHA384",
        "TLS_CHACHA20_POLY1305_SHA256"
    ]

    sniStrict = true

  [tls.options.mintls13]
    minVersion = "VersionTLS13"

[http.middlewares]
  [http.middlewares.compress-all.compress]

  [http.middlewares.https-only.redirectScheme]
    scheme = "https"
    permanent = true

  [http.middlewares.security-headers.headers]
    BrowserXssFilter = true
    ContentTypeNosniff = true
    ForceSTSHeader = true
    FrameDeny = true
    SSLRedirect = true
    STSIncludeSubdomains = true
    STSPreload = true
    STSSeconds = 315360000

  [http.middlewares.strip-prefix.chain]
    middlewares = [ "strip-prefix-1", "strip-prefix-2"]

  [http.middlewares.strip-prefix-1.redirectregex]
    regex="^(https?://[^/]+/[a-z0-9_]+)$"
    replacement="${1}/"
    permanent = true

  [http.middlewares.strip-prefix-2.stripprefixregex]
    regex = [ "/[a-z0-9_]+" ]

```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
(paste your output here)
```

**Comment by trajano** [comment]
I have tested that it does create and forward the X-B3-Traceid header properly by adding the following block

        location /w {
            proxy_set_header Host $host;
            proxy_pass http://whoami/;
        }

**Comment by trajano** [comment]
It appears jaeger has propagation support but not zipkin https://github.com/containous/traefik/blob/88ebac942ed8993d2ffa630efd7e422976c7e4f4/docs/content/observability/tracing/jaeger.md

**Comment by jcchavezs** [comment]
Since propagation is being done by the underlaying opentracing interfaces propagation should work the same for both. I am up to help more on this.

**Comment by wdalmut** [comment]
i have the same problem with zipkin and different traefik proxies in front of group of microservices

```
                         +---------------+           +---------------+
                         |               |           |               |
                         |    Traefik    +----------->   Service A   |
                         |               |           |               |
                         +------+--------+           +---------------+
                                ^
                                |
                                +----------------------------+
                                                             |
+----------------+       +---------------+           +-------+-------+
|                |       |               |           |               |
|  User Request  +------>+    Traefik    +----------->   Service B   |
|                |       |               |           |               |
+----------------+       +---------------+           +---------------+

```

Tracing information from "Service B" to "Service A" via a new Traefik is completely lost and appears as a new trace in Zipkin.

**Comment by jcchavezs** [comment]
I will check it ASAP

søn. 17. nov. 2019, 14:09 skrev Walter Dal Mut <notifications@github.com>:

**Comment by wdalmut** [comment]
@jcchavezs i am reviewing different things about my application. I discovered some bugs in my app instrumentation, so it could not a traefik bug this propagation issue. I will check different things in my codebase and i will report my test result. Sorry if this is not a traefik bug.

**Comment by trajano** [comment]
When I OP this, I was using traefik to nginx, but it also appears to occur when I do Traefik to traefik https://github.com/trajano/trajano-swarm

**Comment by jcchavezs** [comment]
Sorry for the delay. I am checking it now.

**Comment by jcchavezs** [comment]
I checked the code and all the potential sources of the problem but I could not find anything burning so I have two hypothesis:
1. First one is that the nginx extension does not support zipkin v2 format (only v1) and hence the trace gets in a weird state. If someone can paste here an example trace (you con export the trace as a json in zipkin).
2. Is everyone using `b3` propagation?

I really want to get this sorted so if you also prefer, you can ping me on https://gitter.im/jcchavezs

**Comment by jcchavezs** [comment]
I found the problem. I opened a PR with the fix: https://github.com/openzipkin-contrib/zipkin-go-opentracing/pull/155

In the other hand, this will also be fixed if we use the `opentracing.HTTPHeadersCarrier` instead of a custom one since [this PR](https://github.com/containous/traefik/blob/master/pkg/tracing/carrier.go#L6) is merged.

Ping @trajano @dduportal @wdalmut

**Comment by wdalmut** [comment]
good i had fixed my instrumentation error but sadly tracing remains wrong... Now i will generate a structure that replicate my applications to expose the problem. i will update this.

**Comment by traefiker** [comment]
Closed by #5920.
